crates/matrix-sdk/src/room/timeline/inner.rs

```diff
@@ -368,12 +370,93 @@ impl<P: ProfileProvider> TimelineInner<P> {
             .await;
         }
     }
+
+    fn update_event_item(&self, index: usize, event_item: EventTimelineItem) {
+        self.items.lock_mut().set_cloned(index, Arc::new(TimelineItem::Event(event_item)))
+    }
 }
 
 impl TimelineInner {
     pub(super) fn room(&self) -> &room::Common {
         &self.profile_provider
     }
+
+    pub(super) async fn fetch_in_reply_to_details(
+        &self,
+        index: usize,
+        mut item: RemoteEventTimelineItem,
+    ) -> Result<RemoteEventTimelineItem> {
+        let TimelineItemContent::Message(message) = item.content.clone() else {
+            return Ok(item);
+        };
+        let Some(in_reply_to) = message.in_reply_to() else {
+            return Ok(item);
+        };
+
+        let details =
+            self.fetch_replied_to_event(index, &item, &message, &in_reply_to.event_id).await;
+
+        // We need to be sure to have the latest position of the event as it might have
+        // changed while waiting for the request.
+        let (index, _) = rfind_event_by_id(&self.items(), &item.event_id)
+            .ok_or(super::Error::RemoteEventNotInTimeline)?;
+
+        item = item.with_content(TimelineItemContent::Message(message.with_in_reply_to(
+            InReplyToDetails { event_id: in_reply_to.event_id.clone(), details },
+        )));
+        self.update_event_item(index, item.clone().into());
+
+        Ok(item)
+    }
+
+    async fn fetch_replied_to_event(
+        &self,
+        index: usize,
+        item: &RemoteEventTimelineItem,
+        message: &Message,
+        in_reply_to: &EventId,
+    ) -> TimelineDetails<Box<RepliedToEvent>> {
+        if let Some((_, item)) = rfind_event_by_id(&self.items(), in_reply_to) {
+            match item.content() {
+                TimelineItemContent::Message(message) => {
+                    return TimelineDetails::Ready(Box::new(RepliedToEvent {
+                        message: message.clone(),
+                        sender: item.sender().to_owned(),
+                        sender_profile: item.sender_profile().clone(),
+                    }));
+                }
+                _ => {
+                    return TimelineDetails::Error(Arc::new(super::Error::UnsupportedEvent.into()))
+                }
+            }
```

Comment:
I think it would be easier to read at a glance that this `if let Some()` always returns early if there was just a single `return` at the end:

```suggestion
            let details = match item.content() {
                TimelineItemContent::Message(message) => {
                    TimelineDetails::Ready(Box::new(RepliedToEvent {
                        message: message.clone(),
                        sender: item.sender().to_owned(),
                        sender_profile: item.sender_profile().clone(),
                    }))
                }
                _ => {
                    TimelineDetails::Error(Arc::new(super::Error::UnsupportedEvent.into()))
                }
            };
            return details;
```